### PR TITLE
Proposal: rework top-level API

### PR DESCRIPTION
**Pull request by JosefWN**
This is an attempt to fix:
- https://github.com/shamblett/coap/issues/75
- https://github.com/shamblett/coap/issues/72
- https://github.com/shamblett/coap/issues/41

My work was loosely inspired by https://pub.dev/packages/http but with an adaptation to how CoAP clients commonly work in other languages (at least those I like 😅). See the examples to get a feeling for things. TLDR: this PR aims to make the library more native to Dart, to better leverage its strengths. Hopefully I'll manage to make a good case for it below...

Unsure:
- Removed the `validate` method on the client, I figured either the user builds his request (for example for etag validation, using `client.send`), or the user relies on the client for building the entire request (for example `client.get('hello')`). This felt like a bit of a hybrid.

Major stuff:
- There are request builders for the most common verbs (`get(...)`, `post(...)` etc.), a more manual `send(...)` for advanced requests and its counterpart `observe(...)` for streaming, which I suppose is not used often enough to warrant a request builder?
- Tightened up the API by exposing less internals and using the `meta` package with its `@protected` annotation for those I couldn't hide completely (will give static analysis warnings). It's now much harder for the user to "do the wrong things".
  - This also reduces the need for some of the error handling that was in place. In fact, we would have needed even more error handling with the old solution, there were plenty of ways to make the client misbehave/crash.
- Major removals:
  - Logging, which is not typically done in libraries (see Dart standard library for example). While the logging helps developers of this library debug, it complicates the code a fair bit, and is an obstacle for users of the library (in fact, all issues mentioned above are touching upon logging and error handling, directly or indirectly). If anything I would suggest adding more integration tests (examples) and working out as many bugs as possible that way. With a really stable API I don't think extensive logging is necessary. I could actually find the issues while working now, with a couple of temporary `print()` calls.
  - The CoAP.NET project hasn't seen active development since mid-2014 (8 years ago!), and keeping binaries in git is generally discouraged. It's also hard to use these tests on a non-Windows machine. In addition to coap.me, there is Californium, a project which seems to be alive and well, and whose plugtest-server you can run yourself cross-platform. If you miss any tests I propose adding more rather than bringing back the testserver directory, which also contained some duplicate code.
  - The flutter example project felt a bit awkward to keep in a Dart library, only more code to maintain. Flutter users will likely find using the library straight-forward anyhow (it was no problem for me at all). 
  - `CoapIChannel` and `CoapIDeliverer` and their implementations were seemingly not serving any purpose, even less so after I went from passing the response back to the request, to using the event bus to communicate. This in turn removes the need for the `CoapNetworkManager` since each client now only has one endpoint (less risk of API user mistakes), as well as the setting `poolUdpConnectionsByClient`. The pooling doesn't have any drawbacks as far as I can see, the more separation the better?
- Now supports async requests over a shared socket. This required separating different components to a greater degree, and relying less on global/static variables. Two major benefits:
1. The code doesn't crash on:
```dart
client.get('large').then((CoapResponse resp) => print(resp));
```
You can even run multiple observe and different requests concurrently (see `examples/get_observe_async.dart`).

2. Requests can be batched at far higher efficiency, example:
```bash
dart sync_vs_async.dart
Sending 10 async requests...
10 async requests took 49 ms
Sending 10 sync requests...
10 sync requests took 393 ms
```

Minor stuff:
- Fixes an issue with observe (no endpoint set, `_doPrepare` called one line too late)
- Fixes an issue where token is not set on the overarching `block1` requests (need it for matching in the client)
- Merged the `test/issues` folder with `examples`
- Removes `notify` and `error` on `observe`, they were unused arguments from what I could see
- Removes `raw_request_creation.dart` as that would no longer be supported (no need to worry about preparing the request manually)
- Many other small fixes...

Performance vs 3.5.0 (in parentheses), as baseline before all of my changes:
```
- 10 sync requests: 362 ms (410 ms)
- Ping: 93 ms (10033 ms) <- returning on reject now, not on timeout
- Blockwise (block2): 799 ms (846 ms) 
- Blockwise (block1): 248 ms (302 ms)
```

Nothing scientific, just a couple of quick checks to ensure my changes that went out in v3.6.0 and in this PR are not making things slower.

### CODE REVIEW

**Comment by JosefWN** [comment]
Please don't merge yet, but feel free to come with input! I'm thinking of moving `send` and `waitForResponse` into the client in a separate commit if it pans out well, which can also be reverted if you guys don't like it, will do it tonight or tomorrow.

**Comment by JosefWN** [comment]
Ok! Ready for review @shamblett, @JKRhb! Linting, formatting & tests are passing, examples are working :)

Not sure how to go about reviewing what turned out to be a gargantuan PR, sorry for that. Perhaps the best is to just start with viewing and running the examples to get a feel for the API. Everything is tested except for multicast, didn't have a good way of testing it.

**Comment by JKRhb** [comment]
Thank you for your hard work, @JosefWN!

> Not sure how to go about reviewing what turned out to be a gargantuan PR, sorry for that.

Maybe aspects like the test server and flutter example removal could be spun off into its own PR? That would reduce the size of the PR a bit

File: examples/multi_client.dart
Comment (JKRhb) [inline examples/multi_client.dart:32]: I noticed one typo here:

```suggestion
    response = await client2.get('test');
```

File: lib/src/network/coap_network_udp.dart
Comment (JKRhb) [inline lib/src/network/coap_network_udp.dart:74]: I noticed the linter is complaining a bit here:

```suggestion
        default:
          break;
```

File: lib/src/coap_request.dart
Comment (JKRhb) [inline lib/src/coap_request.dart:27]:
```dart
class CoapRequest extends CoapMessage {
  /// Initializes a request message.
  /// Defaults to confirmable
  CoapRequest(int code, {confirmable = true})
      : super(
            code: code,
            type: confirmable ? CoapMessageType.con : CoapMessageType.non);

```

File: lib/src/coap_message.dart
Comment (JKRhb) [inline lib/src/coap_message.dart:24]:
```dart
class CoapMessage {
  /// Constructor
  CoapMessage(
      {this.type = CoapMessageType.unknown, this.code = CoapCode.notSet});
```

**Review by JKRhb — COMMENTED** [review]
I think both the `CoapMessage` and the `CoapRequest` constructors could be simplified a bit. (This requires some more changes at other places in the code base.)

File: examples/multi_client.dart
Comment (JosefWN) [reply examples/multi_client.dart]: Ah, whoops, fixed: https://github.com/shamblett/coap/pull/76/commits/c62836d6601d0f0800884d4125628e452cb5502e

File: lib/src/network/coap_network_udp.dart
Comment (JosefWN) [reply lib/src/network/coap_network_udp.dart]: Fixed: https://github.com/shamblett/coap/pull/76/commits/c62836d6601d0f0800884d4125628e452cb5502e

File: lib/src/coap_request.dart
Comment (JosefWN) [reply lib/src/coap_request.dart]: Much cleaner! Changed: https://github.com/shamblett/coap/pull/76/commits/c62836d6601d0f0800884d4125628e452cb5502e

File: lib/src/coap_message.dart
Comment (JosefWN) [reply lib/src/coap_message.dart]: Agreed: https://github.com/shamblett/coap/pull/76/commits/c62836d6601d0f0800884d4125628e452cb5502e

**Comment by JosefWN** [comment]
> Maybe aspects like the test server and flutter example removal could be spun off into its own PR? That would reduce the size of the PR a bit

Yeah, the reason I included them in the same PR was because I didn't want to invest time in cleaning them up and fixing potential changes causing them to break on master. Going out now but I can re-add them tomorrow and then make a new PR if/once this one gets merged.

**Comment by JosefWN** [comment]
@JKRhb done, also reduced duplicate code in the client a bit :)

File: lib/src/net/coap_exchange.dart
Comment (JKRhb) [inline lib/src/net/coap_exchange.dart:40]: Hmm, is it intentional that this is not a doc comment anymore?

File: lib/src/net/coap_iendpoint.dart
Comment (JKRhb) [inline lib/src/net/coap_iendpoint.dart:29]:
```suggestion
  /// The namespace which the endpoint belongs to
```

File: lib/src/observe/coap_observe_relation.dart
Comment (JKRhb) [inline lib/src/observe/coap_observe_relation.dart:14]: I think this file could probably be adjusted a bit:

```dart
/*
 * Package : Coap
 * Author : S. Hamblett <steve.hamblett@linux.com>
 * Date   : 31/10/2017
 * Copyright :  S.Hamblett
 */

part of coap;

/// Represents a relation between a client endpoint and a resource on this
/// server.
class CoapObserveRelation {
  /// Constructs a new observe relation.
  ///
  /// Takes the observing [endpoint], the observed [resource], and the
  /// [exchange] that tries to establish the observe relation as arguments.
  CoapObserveRelation(this.cfg, CoapObservingEndpoint endpoint,
      CoapIResource resource, CoapExchange exchange)
      : _endpoint = endpoint,
        _resource = resource,
        _exchange = exchange,
        established = true;

  DefaultCoapConfig cfg;

  final CoapObservingEndpoint _endpoint;

  /// Source endpoint of the observing endpoint
  InternetAddress? get source => _endpoint.endpoint;
  final CoapIResource _resource;

  /// The resource
  CoapIResource? get resource => _resource;
  final CoapExchange _exchange;

  /// The exchange
  CoapExchange get exchange => _exchange;

  /// Current control notification
  CoapResponse? currentControlNotification;

  /// Next control notification
  CoapResponse? nextControlNotification;

  /// Key
  String get key => '$source#${_exchange.request?.tokenString}';

  /// A value indicating if this relation has been established
  bool established;
  DateTime _interestCheckTime = DateTime.now();
  int _interestCheckCounter = 1;

  /// The notifications that have been sent, so they can be
  /// removed from the Matcher.
  final Queue<CoapResponse?> _notifications = Queue<CoapResponse?>();

  /// Cancel this observe relation.
  void cancel() {
    // Stop ongoing retransmissions
    if (_exchange.response != null) {
      _exchange.response!.isCancelled = true;
    }
    established = false;
    _resource.removeObserveRelation(this);
    _endpoint.removeObserveRelation(this);
    _exchange.complete = true;
  }

  /// Cancel all observer relations that this server has
  /// established with this's realtion's endpoint.
  void cancelAll() {
    _endpoint.cancelAll();
  }

  /// Notifies the observing endpoint that the resource has been changed.
  void notifyObservers() {
    // Makes the resource process the same request again
    _resource.handleRequest(_exchange);
  }

  /// Check
  bool check() {
    var check = false;
    final now = DateTime.now();
    check = check ||
        _interestCheckTime
            .add(Duration(milliseconds: cfg.notificationCheckIntervalTime))
            .isBefore(now);
    check = check ||
        (++_interestCheckCounter >= cfg.notificationCheckIntervalCount);
    if (check) {
      _interestCheckTime = now;
      _interestCheckCounter = 0;
    }
    return check;
  }

  /// Add a notification
  void addNotification(CoapResponse notification) {
    _notifications.add(notification);
  }

  /// Clear notifications
  Iterable<CoapResponse?> clearNotifications() {
    Iterable<CoapResponse?> list;
    list = _notifications.toList();
    _notifications.clear();
    return list;
  }
}
```

File: lib/src/observe/coap_observe_relation.dart
Comment (JKRhb) [inline lib/src/observe/coap_observe_relation.dart:25]: `cfg` could be renamed to `config`

File: lib/src/stack/coap_blockwise_layer.dart
Comment (JKRhb) [inline lib/src/stack/coap_blockwise_layer.dart:264]: I think the type of `currentNUM` could be changed from `int?` to `int`.

File: lib/src/stack/coap_chain.dart
Comment (JKRhb) [inline lib/src/stack/coap_chain.dart:268]: I think the `!` could be removed here

```suggestion
    return e as CoapIEntry<TFilter, TNextFilter>;
```

File: lib/src/stack/coap_chain.dart
Comment (JKRhb) [inline lib/src/stack/coap_chain.dart:278]: Same here I think.

File: lib/src/stack/coap_chain.dart
Comment (JKRhb) [inline lib/src/stack/coap_chain.dart:424]: I think this null check operator is also not needed.

**Review by JKRhb — COMMENTED** [review]
I think the PR looks pretty good, see a last couple of comments/suggestions below :)

File: lib/src/stack/coap_chain.dart
Comment (JosefWN) [reply lib/src/stack/coap_chain.dart:268]: No: https://dart-lang.github.io/linter/lints/cast_nullable_to_non_nullable.html

Comment (JKRhb) [reply lib/src/stack/coap_chain.dart:268]: Oh, I see, thanks for the clarification!

File: lib/src/stack/coap_chain.dart
Comment (JosefWN) [reply lib/src/stack/coap_chain.dart:424]: I think in general it's better to be explicit in the code. In this case the cost is also only one character to clarify that the variable can be null, but that we are not expecting it to be?

File: lib/src/coap_client.dart
Comment (JosefWN) [reply lib/src/coap_client.dart:71]: Ah, we should remove the milliseconds thing altogether in the comments there, since now you can just set `client.timeout = Duration(seconds: 2)` for example, which makes it explicit to the library user anyway. The unit is only important for the `CoapConfig` now.

File: lib/builder.dart
Comment (JosefWN) [reply lib/builder.dart:62]: Yeah, in general I was planning on making a pass over the code checking for that as well. There are numerous places in the code where we could be using final.

We could also consider adding some basic linter rules, for example:
```yaml
linter:
  rules:
    - prefer_final_fields
    - prefer_final_in_for_each
    - prefer_final_locals
    - prefer_final_parameters
    - prefer_const_constructors
    - prefer_const_constructors_in_immutables
    - prefer_const_declarations
    - prefer_const_literals_to_create_immutables
```

The base linter config is pretty lenient.

File: lib/src/stack/coap_chain.dart
Comment (JKRhb) [reply lib/src/stack/coap_chain.dart:424]: Yeah, also considering your last comment, it is certainly better to leave it as for now :)

File: lib/src/net/coap_exchange.dart
Comment (JosefWN) [reply lib/src/net/coap_exchange.dart]: Not intentional, fixed: https://github.com/shamblett/coap/pull/76/commits/bf2e194af342174accb5b8c413397e7d3f129cdf

File: lib/src/net/coap_iendpoint.dart
Comment (JosefWN) [reply lib/src/net/coap_iendpoint.dart]: Fixed: https://github.com/shamblett/coap/pull/76/commits/bf2e194af342174accb5b8c413397e7d3f129cdf

File: lib/src/observe/coap_observe_relation.dart
Comment (JosefWN) [reply lib/src/observe/coap_observe_relation.dart]: Fixed: https://github.com/shamblett/coap/pull/76/commits/ad10f99673d9303bd2709d6c05a78c065909d09b

File: lib/src/stack/coap_chain.dart
Comment (JosefWN) [reply lib/src/stack/coap_chain.dart:424]: Ok! Marking this as resolved now then :)

File: lib/src/coap_client.dart
Comment (JosefWN) [reply lib/src/coap_client.dart:107]: Yeah, I've seen enum used in other libs for SZX, just like it is specified in the RFC. Makes it more transparent, now the payload size will be rounded and unless the user has paid attention to the doc comment they might be a bit surprised. I can change to an enum and have a look at a block1 setting as well.

File: lib/src/coap_client.dart
Comment (JosefWN) [reply lib/src/coap_client.dart:62]: Yes, I don't know how we should treat the config. I figured it could make more sense to move our non-RFC timeout default (which we had apparently set to 32767 ms) there as well, so we have a single config. Then again, we already have an implicit total timeout specified in the config, so should we define yet another timeout for the client like we do now? Regardless, the config will likely span multiple RFC's (7252 for UDP, 8232 for TCP, 7641 for observe, 7959 for blockwise), so it will probably not map 1:1 to RFC 7252.

File: lib/src/coap_client.dart
Comment (JosefWN) [reply lib/src/coap_client.dart:75]: Yeah, the reason I did it this way was to avoid repeated type conversions, but I can change, I'm not strongly in favor of either approach and it's a tiny cost.

File: lib/src/coap_client.dart
Comment (JosefWN) [reply lib/src/coap_client.dart]: Fixed: https://github.com/shamblett/coap/pull/76/commits/23f12832be6d414080a70c4193c1a4afe34c44e9

File: lib/src/coap_client.dart
Comment (JosefWN) [reply lib/src/coap_client.dart]: Fixed: https://github.com/shamblett/coap/pull/76/commits/23f12832be6d414080a70c4193c1a4afe34c44e9

File: lib/src/coap_client.dart
Comment (JosefWN) [reply lib/src/coap_client.dart]: Fixed the constructor: https://github.com/shamblett/coap/pull/76/commits/23f12832be6d414080a70c4193c1a4afe34c44e9

If you want I can remove `defaultTimeout` from the config and add a separate issue to discuss whether we should specify the timeout twice?

File: lib/src/observe/coap_observe_relation.dart
Comment (JosefWN) [reply lib/src/observe/coap_observe_relation.dart]: Fixed: https://github.com/shamblett/coap/pull/76/commits/806d2aabc363905fe1dde8efcbf881d42f3118a3

File: lib/src/coap_client.dart
Comment (JKRhb) [reply lib/src/coap_client.dart]: > Fixed the constructor: [23f1283](https://github.com/shamblett/coap/commit/23f12832be6d414080a70c4193c1a4afe34c44e9)
> 
> If you want I can remove `defaultTimeout` from the config and add a separate issue to discuss whether we should specify the timeout twice?

That sounds like a good idea :)

Comment (JosefWN) [reply lib/src/coap_client.dart]: Fixed: https://github.com/shamblett/coap/pull/76/commits/175151b7968921eae2d53806b700e69a345a19d5

File: lib/src/stack/coap_blockwise_layer.dart
Comment (JosefWN) [reply lib/src/stack/coap_blockwise_layer.dart]: Fixed: https://github.com/shamblett/coap/pull/76/commits/e3d2bff4bf1ebcf93e8a832e13d29aa9c4a2f3f7

File: lib/src/coap_client.dart
Comment (JosefWN) [reply lib/src/coap_client.dart:106]: Also I think all block1 requests use early negotiation always (since the first block will contain data anyway), but I could be wrong: https://datatracker.ietf.org/doc/html/rfc7959

Comment (JKRhb) [reply lib/src/coap_client.dart:106]: > I suggest holding off until we have enhanced enums in Dart 2.17, created issue: https://github.com/shamblett/coap/issues/80

That sounds reasonable :)

> Also I think all block1 requests use early negotiation always (since the first block will contain data anyway), but I could be wrong: https://datatracker.ietf.org/doc/html/rfc7959

Hmm, I think by indicating the block1 value you could set the client's preferred/initial block size (c. f. figure 9). Having another look into the source code, though, I am not sure if "manually" adding a block1 option/setting the size would actually be processed at the moment. Maybe this is another potential issue that can be addressed in a follow-up PR :)

Comment (JosefWN) [reply lib/src/coap_client.dart:106]: Yeah, but what makes the negotiation different is that for every block in a block1 request we set:
```dart
    block
        .addOption(CoapBlockOption.fromParts(optionTypeBlock1, num, szx, m: m));
```

This includes the first request/block. There will still be a negotiation, where the block1 size in subsequent messages is adjusted to the smallest of the client's preferred block1 size and the server's maximum block1 size.

In a block2 get for example, the get request itself is not blockwise, so we can negotiate early by including the block2 options as an instruction for the blockwise response from the server.

Comment (JosefWN) [reply lib/src/coap_client.dart:106]: But in general I'm wondering if we should include this option to alter block size for specific requests at all, I just kept it because it was exposed to the user in the first place.

The block size is also set depending on connectivity (likelihood of IP packet fragmentation and such), so it would make sense to use the same settings for the whole client since it now maps 1:1 with a specific server, and this is specified in the config already. If you want to push larger payloads to the client you wait longer, you typically don't increase the block size.

Comment (JosefWN) [reply lib/src/coap_client.dart:107]: Example: https://github.com/shamblett/coap/pull/76/commits/8dc6fb302e08398a7324e25ab3966f75806e0d66

So there is no `defaultBlockSize` but a `preferredBlockSize` as per the RFC. This is what the client as a whole prefers. Then you can choose whether or not to instruct the server to use the client's preference in its block2 response.

The whole negotiation is "piggybacked" with the requests/responses, so for block1 you can't negotiate early unless you want an extra roundtrip (very expensive), because the first block1 request already contains a payload.

Comment (JKRhb) [reply lib/src/coap_client.dart]: Oh, that is a very elegant solution :) Very cool, thank you! :)

**Review by JKRhb — APPROVED** [review]
LGTM :) @shamblett Can you also have a look on this PR? :)

File: lib/src/coap_client.dart
Comment (JosefWN) [reply lib/src/coap_client.dart]: Thank you, too, for focusing on the uglier parts of my refactor to make it better :)

File: lib/src/net/coap_exchange.dart
Comment (JKRhb) [inline lib/src/net/coap_exchange.dart:230]: I guess here we could use something like this?

```suggestion
    _hash = Object.hash(_uri, _endpoint);
```

File: lib/src/coap_option.dart
Comment (JosefWN) [reply lib/src/coap_option.dart:183]: Ah, nice catch. In addition, if all fields are included in `Object.hash()`, I don't think we need to override `hashCode` or `operator ==`?

Comment (JosefWN) [reply lib/src/coap_option.dart]: Okay, I got found a couple more places to clean up while I was at it: https://github.com/shamblett/coap/pull/76/commits/4cf829c056e6a63e408a3e93d991d548a413952f

With client mapping 1:1 to a socket we don't need `CoapKeyUri`, we can match blockwise requests on token. Having a `List<Map<int?, CoapExchange>>` for matching on id and a `Map<String, CoapExchange>` for matching on token works just fine. There is no need for the other two wrapping classes.

Re-ran all examples and everything works.

**Comment by shamblett** [comment]
I've merged this to its own branch for now, issue75, while I check it out.

One question about logging, the original logging was really supplied so users could just turn on logging and supply me with a log of what happened when they raised an issue, without this we are pretty blind. The old logging although as you say problematic in itself did allow post mortem debugging for issue investigation. Are we now saying we are just replacing the old logging with print statements? This is OK but we mustn't lose detail here, the examples now provide minimal logging(OK these of course work so that maybe why). 

Will we still be able to see message detail, i.e token's/sequence numbers/flags/block counts etc. that we used to do?
These may be invaluable,especially in cases where the CoAP server the user is using doesn't work properly(we have had this in the past). We may not be able to reproduce the fault, nor give enough detail to the user as to what the problem is.

Sorry if I've missed something here.

**Comment by JosefWN** [comment]
Hi @shamblett, and sorry for the large PR!

Personally I'm a big fan of the event bus since it helps us not only decouple the components within the library, but also gives us the opportunity to introspect the request/response flow. In this PR I always fire events (even if there is a hook) for this purpose. I'm also always including data with all events, so we can differentiate between multiple concurrent requests.

To make debug logging easier, we could expose the bus on the client as well, and add `toString` methods on the events. With that in place, I think what you are asking for could be achieved with this one-liner:
```dart
client.events.on().listen(print);
```

Example: https://github.com/JosefWN/coap/commit/087aafb43827b91d675001f7b15fd5d10b50dbc1

If you try running the `log_events.dart` example I included in the commit, you will see all blockwise activity.

This also makes it possible for library users to listen for and act on any events they choose, without imposing a specific logger or event handler on them (of course, we would still handle the events internally as required, but we wouldn't programmatically obscure anything from the library user). If we are missing important events for debugging we can just add more of them. What do you think?

Another thing we could do when users encounter issues is to ask them to provide us with a minimal example reproducing that issue. Shifting the burden of proof to the person with the problem will reduce the debugging load on us even further, as we could just run their problematic example code, or a very similar piece of code, against another server.

**Comment by shamblett** [comment]
This sounds good to me, the existing logging did need to go, I'll have a look at log_events.dart but I'm happy with this approach.

**Pull request merged**
